src/main/java/dev/mccue/resolve/graph/DependencyTree.java

```diff
@@ -1,20 +1,86 @@
 package dev.mccue.resolve.graph;
 
 import dev.mccue.resolve.core.Dependency;
-import dev.mccue.resolve.doc.Coursier;
-import dev.mccue.resolve.doc.Incomplete;
 
-import java.util.List;
+import java.util.*;
+import java.util.stream.Collectors;
 
-@Incomplete
-@Coursier("https://github.com/coursier/coursier/blob/f5f0870/modules/core/shared/src/main/scala/coursier/graph/DependencyTree.scala")
-public interface DependencyTree {
-    Dependency dependency();
+public sealed interface DependencyTree {
 
-    boolean excluded();
+    static List<DependencyTree> of(
+            Resolution resolution,
+            List<Dependency> roots,
+            boolean withExclusions
+    ) {
+        List<Dependency> initialRoots;
+        if (!roots.isEmpty()) {
+            initialRoots = resolution.rootDependencies;
+        } else {
+            initialRoots = roots;
+        }
 
-    String reconciledVersion();
-    String retainedVersion();
+        return initialRoots.stream().<DependencyTree>map(dep -> new Node(dep, false, resolution, withExclusions)).toList();
+    }
 
-    List<DependencyTree> children();
+    record Node(
+            Dependency dependency,
+            Boolean excluded,
+            Resolution resolution,
+            Boolean withExclusions
+    ) implements DependencyTree {
+        String reconciledVersion() {
+            if (resolution.reconciledVersion.dependency.module() != null) {
+                return resolution.reconciledVersion.dependency.module().toString();
+            } else {
+                return resolution.reconciledVersion.dependency.version();
+            }
+        }
+
+        String retainedVersion() {
+            if (resolution.retainedVersion.dependency.module() != null) {
+                return resolution.retainedVersion.dependency.module().toString();
+            } else {
+                return resolution.retainedVersion.dependency.version();
+            }
+        }
+
+        List<DependencyTree> children() {
+            if (excluded) {
+                return null;
```

Comment:
Strange this returns null - idk the context here. Why not empty list?

Reply:
Yeah, that's how code was written in Scala.
```
// don't make that a val!! issues with cyclic dependencies
    // (see e.g. edu.illinois.cs.cogcomp:illinois-pos in the tests)
    def children: Seq[DependencyTree] =
      if (excluded)
        Nil
      else {
        val dep0 = dependency.withVersion(retainedVersion)
```

Reply:
Maybe the return should be `Optional<List<DependencyTree>>` then? idk